src/apps/INTSC_HEXHEX.cpp

```diff
@@ -0,0 +1,309 @@
+  Index_type n_std_intsc  = getActualProblemSize() ;
+  Index_type n_subz_intsc = 8L * n_std_intsc ;
+
+  // coordinates for donor zone
+  Real_type xdzone[8] =
+      { m_xmin, m_xmax, m_xmin, m_xmax, m_xmin, m_xmax, m_xmin, m_xmax } ;
+
+  Real_type ydzone[8] =
+      { m_ymin, m_ymin, m_ymax, m_ymax, m_ymin, m_ymin, m_ymax, m_ymax } ;
+
+  Real_type zdzone[8] =
+      { m_zmin, m_zmin, m_zmin, m_zmin, m_zmax, m_zmax, m_zmax, m_zmax } ;
+
+  Real_type xtzone[8], ytzone[8], ztzone[8] ;
+  for ( Index_type i=0 ; i<8 ; ++i ) {
+    xtzone[i] = xdzone[i] + m_shift ;
+    ytzone[i] = ydzone[i] + m_shift ;
+    ztzone[i] = zdzone[i] + m_shift ;
+  }
+
+  Real_ptr dcoord ;   // donor  coordinates [24]
+  Real_ptr tcoord ;   // target coordinates [24]
+
+  // Expanded donor and target coordinates on host
+  Real_ptr ds_h, ts_h ;
+
+  do {
+    allocDataForInit ( dcoord, 24, Base_Seq ) ;
+    memcpy ( dcoord   , xdzone, 8*sizeof(Real_type) ) ;
+    memcpy ( dcoord+ 8, ydzone, 8*sizeof(Real_type) ) ;
+    memcpy ( dcoord+16, zdzone, 8*sizeof(Real_type) ) ;
+
+    allocDataForInit ( tcoord, 24, Base_Seq ) ;
+    memcpy ( tcoord   , xtzone, 8*sizeof(Real_type) ) ;
+    memcpy ( tcoord+ 8, ytzone, 8*sizeof(Real_type) ) ;
+    memcpy ( tcoord+16, ztzone, 8*sizeof(Real_type) ) ;
+
+    allocDataForInit ( m_vv, 4L*n_subz_intsc, Base_Seq ) ;
+
+    allocDataForInit ( ds_h, 24L*n_subz_intsc, Base_Seq ) ;
+    allocDataForInit ( ts_h, 24L*n_subz_intsc, Base_Seq ) ;
+
+  } while ( false ) ;
+
+  //  Repeat the same calculation n_subz_intsc times, expand the
+  //  same donor and target zones.
+  for ( Index_type k=0 ; k < n_subz_intsc ; ++k ) {
+    memcpy ( ds_h + 24L*k, dcoord, 24*sizeof(Real_type) ) ;
+    memcpy ( ts_h + 24L*k, tcoord, 24*sizeof(Real_type) ) ;
+  }
+
+  allocAndCopyHostData ( m_dsubz, ds_h, 24L*n_subz_intsc, vid ) ;
+  allocAndCopyHostData ( m_tsubz, ts_h, 24L*n_subz_intsc, vid ) ;
+
+  do {
+    deallocData ( ds_h, Base_Seq ) ;
+    deallocData ( ts_h, Base_Seq ) ;
+    deallocData ( dcoord, Base_Seq ) ;
+    deallocData ( tcoord, Base_Seq ) ;
+  } while ( false ) ;
+
+  const Int_type block_size = default_gpu_block_size ;
+  m_nthreads = 72L * n_subz_intsc ;
+  m_gsize    = RAJA_DIVIDE_CEILING_INT(m_nthreads, block_size) ;
+
+  // intermediate volumes, moments
+  allocData ( m_vv_int, 8L*m_gsize, vid ) ;
+
+  allocAndInitDataConst ( m_vv_out, 4L*n_subz_intsc, 0.0, vid ) ;
+}
+
+
+//   Number of subzone intersections = 8 * number of standard intersections.
+//
+void INTSC_HEXHEX::check_intsc_volume_moments
+    ( Index_type const n_subz_intsc,  // number of subzone intersections
+      Real_const_ptr vv,   // computed volumes, moments on the host
+      VariantID vid )   // Print variant name in case of error
+{
+  Int_type rank = 0;
+#if defined(RAJA_PERFSUITE_ENABLE_MPI)
+  MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+#endif // RAJA_PERFSUITE_ENABLE_MPI
+
+  // Check on rank 0, other ranks are identical.
+  if ( rank == 0 ) {
```

Comment:
I would prefer to simplify this code a bit and let all ranks do this, then only rank 0 will actually print due to the implementation of getCout.